week6/week6_prepartion_database/Program.cs

```diff
@@ -0,0 +1,271 @@
+        while (await reader.ReadAsync())
+        {
+          var user = reader.GetString(reader.GetOrdinal("user"));
+          var host = reader.GetString(reader.GetOrdinal("host"));
+          users.Add(new User(user, host));
+        }
+        return Results.Ok(users);
+      }
+    }
+  }
+});
+
+/*
+< summary >
+Endpoint for getgetting users. starting from C# 8 it no longer need inner {}
+</summary>
+*/
+app.MapGet("/users2", async (IConfiguration configuration) =>
+{
+  // starting from C# 8 you no longer need inner {}
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var users = await connection.QueryAsync<User>("SELECT * FROM mysql.user");
+  return Results.Ok(users);
+});
+
+/*
+< summary >
+Endpoint for getgetting users
+</summary>
+*/
+app.MapGet("/products", async (IConfiguration configuration) =>
+{
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var products = await connection.QueryAsync<Product>("SELECT id, name, price,description FROM dapper.products");
+  return Results.Ok(products);
+});
+
+
+/*
+<question>
+    Implement the endpoint /product/{id}that:
+    1.Changes the SQL to add WHERE id = {id}
+    2.Instead of a list, returns .FirstOrDefault
+    3.Bonus: if there is no such user, the list would be empty. What should REST API return when you want a resource that doesn’t exist? E.g. you call it with /product/666?
+    4.Bonus: if the user passes negative number e.g. -1, what should REST API return in that case?
+</questions>
+*/
+app.MapGet("/products/{id}", async (IConfiguration configuration, int id) =>
+{
+  if (id < 0)
+  {
+    //#####
+    throw new Exception("id can not be negative");//??
+    // Results.BadRequest($" you entered id:{id},id can not be negative"); 
+  }
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var products = await connection.QueryAsync<Product>($"SELECT id, name, price,description FROM dapper.products where id={id}");
+  if (products.Count() == 0)
+  {
+    return Results.NotFound($" The  product with id:{id} does not exists in the database !!");
+  }
+  return Results.Ok(products);
+});
+
+/*Instead of a list, returns .FirstOrDefault*/
+app.MapGet("/firstordefault/{id}", async (IConfiguration configuration, int id) =>
+{
+  if (id < 0)
+  {
+    throw new Exception("id can not be negative");
+    // Results.BadRequest($" you entered id:{id},id can not be negative");
+  }
+  //#####
+  // how to checko if the input is letter instead of number or is not a number???
+  // var isNumeric = int.TryParse((string)id, out int value);
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var products = await connection.QueryAsync<Product>($"SELECT id, name, price,description FROM dapper.products where id={id}");
+  if (products.Count() == 0)
+  {
+    throw new Exception("product id not found");
```

Comment:
What do you expect from an endpoint if you ask for something and it is not found?

Reply:
it should return notfound !  but 
when i use Results.notfound("id not found"), it gives error ...  how to fix this ?

Reply:
ok i got the point. it works now when i use Results. ...  in all return of the endpoint ! thank you